eureka2-client/src/main/java/com/netflix/eureka2/client/resolver/RoundRobinServerResolver.java

```diff
@@ -0,0 +1,143 @@
+
+    private final int warmUpTimeout;
+    private final TimeUnit timeUnit;
+
+    private final Observable<ChangeNotification<Server>> serverSource;
+    private final AtomicReference<List<Server>> serverCacheRef;
+    private final AtomicInteger positionRef;
+
+    protected RoundRobinServerResolver(Server... servers) {
+        this(sourceFromList(servers), 10, TimeUnit.SECONDS);
+    }
+
+    protected RoundRobinServerResolver(Observable<ChangeNotification<Server>> serverSource) {
+        this(serverSource, 10, TimeUnit.SECONDS);
+    }
+
+    protected RoundRobinServerResolver(Observable<ChangeNotification<Server>> serverSource, int warmUpTimeout, TimeUnit timeUnit) {
+        this.serverSource = serverSource;
+        this.warmUpTimeout = warmUpTimeout;
+        this.timeUnit = timeUnit;
+        this.serverCacheRef = new AtomicReference<List<Server>>(new ArrayList<Server>());
+        this.positionRef = new AtomicInteger(new Random().nextInt(100));
+    }
+
+    public RoundRobinServerResolver withWarmUpConfiguration(int newWarmUpTimeout, TimeUnit newTimeUnit) {
+        return new RoundRobinServerResolver(serverSource, newWarmUpTimeout, newTimeUnit);
+    }
+
+    @Override
+    public void close() {
+        serverCacheRef.get().clear();
+    }
+
+    @Override
+    public Observable<Server> resolve() {
+        return refreshServerCache().concatMap(new Func1<List<Server>, Observable<? extends Server>>() {
+            @Override
+            public Observable<? extends Server> call(List<Server> servers) {
+                if (servers.isEmpty()) {
+                    return Observable.error(SERVER_CACHE_EMPTY_EXCEPTION);
+                }
+
+                int currentPos = positionRef.getAndIncrement();
+                Server toReturn = servers.get(currentPos % servers.size());
+                return Observable.just(toReturn);
+            }
+        });
+    }
+
+    /**
+     * Connect and warm up the serverCache to the server source. onComplete the result Observable once the
+     * serverCache has warmed up.
+     */
+    private Observable<List<Server>> refreshServerCache() {
+        return serverSource
+                .compose(ChangeNotifications.<Server>buffers())
+                .compose(ChangeNotifications.<Server>snapshots())
+                .materialize()
+                .concatMap(new Func1<Notification<LinkedHashSet<Server>>, Observable<? extends List<Server>>>() {
+                    @Override
+                    public Observable<? extends List<Server>> call(Notification<LinkedHashSet<Server>> rxNotification) {
+                        switch (rxNotification.getKind()) {
+                            case OnNext:
+                                LinkedHashSet<Server> servers = rxNotification.getValue();
+                                if (servers.isEmpty()) {  // if onNext is empty, do nothing and wait
+                                    return Observable.never();
```

Comment:
This observable never completes, which will always result in a timeout below. We should use filter to get rid of empty server lists or just emit Observable.empty(). ```take(1)``` below will wait for first non-empty list.